Merge order

tinmarino opened the pull request:
Hi Bduggan, always a pleasure to summit PR here.

This one is not critical but deserves explanations (sorry for making it so long):

* __Problem:__ For my Vim client [jupyter-vim](https://github.com/jupyter-vim/jupyter-vim) the connection I implemented, get some info from the kernel (pid, hostname). This comes from the iostream for Raku. But this one comes too late due to async iostream (old PR for thread io support)

* __Idea:__ So I wanted the `execute_reply` to wait for the stream `idle` was sent to the socket.

* __Solution:__ The solution I chose here (criticable of course) is to use the awesome `Async::Lock`, unlock when 'idle' is sent, they kernel can `execute_reply` and client is sure that the stream is in the socket upto `idle`.


PS 1: The python code: 
 In python, they flush, sleep, https://github.com/ipython/ipykernel/blob/ff122dd62388af07354cc209387dfa271294e8ee/ipykernel/kernelbase.py#L549-L556, flush again, send anything and Pool it. Too many contributors I guess.


```python
# Flush output before sending the reply.
sys.stdout.flush()
sys.stderr.flush()
# FIXME: on rare occasions, the flush doesn't seem to make it to the
# clients... This seems to mitigate the problem, but we definitely need
# to better understand what's going on.
if self._execute_sleep:
    time.sleep(self._execute_sleep)

...

sys.stdout.flush()
sys.stderr.flush()
self._publish_status(u'idle')
# flush to ensure reply is sent
self.control_stream.flush(zmq.POLLOUT)

```

 PS2: A screenshot of the Vim plugin:
![raku-jupyter-order](https://user-images.githubusercontent.com/6123962/78065596-e640fb00-7369-11ea-9f4e-29ad7b0db5af.png)

bduggan commented:
So -- let me see if I understand the issue -- the `execute_input` message on `iopub` needs to be sent _before_ the `execute_reply` message is sent on `shell`?  That makes sense.  Would you mind adding a test for this?  As far as implementation -- it looks good but perhaps `iopub_channel` could just be a supply, since we're just calling `.Supply` on it anyway.

tinmarino commented:
You understood perfectly my point. I agree with your improvements and I added the 3 (safety):

1. Channel -> Supplier: (ea7020c) so a lighter object is shared (i.e. minimal).
2. Test: (85df22d): I made them in end-to-end, it was much easyer. I added an optional parameters (`flags`) to the `receive` function in `Service.rakumod` so a caller can choose to receive without waiting.
3. Replace Lock (low level) with a timeout promise (mush higher level) (efd1db3 ): So if something goes wrong in some thread (which should not happen anyway). After a hard coded 0.2 sec (python chose 0.150), the reply is sent anyway and the loop do not break.

All test are passing on my machine 

```bash
MVM_SPESH_DISABLE=1 P6_JUPYTER_TEST_AUTOCOMPLETE=1 P6_JUPYTER_TEST_END_TO_END=1 zef test --debug  .
```

bduggan commented:
Great!  Looks good
